espnet2/train/preprocessor.py

```diff
@@ -203,85 +203,89 @@ def __init__(
         else:
             self.noises = None
 
+    def _convolve_rir(self, speech, power):
+        rir_path = np.random.choice(self.rirs)
+        rir = None
+        if rir_path is not None:
+            rir, _ = soundfile.read(rir_path, dtype=np.float64, always_2d=True)
+
+            # rir: (Nmic, Time)
+            rir = rir.T
+
+            # speech: (Nmic, Time)
+            # Note that this operation doesn't change the signal length
+            speech = scipy.signal.convolve(speech, rir, mode="full")[
+                :, : speech.shape[1]
+            ]
+            # Reverse mean power to the original power
+            power2 = (speech[detect_non_silence(speech)] ** 2).mean()
+            speech = np.sqrt(power / max(power2, 1e-10)) * speech
+        return speech, rir
+
+    def _add_noise(self, speech, power):
+        nsamples = speech.shape[1]
+        noise_path = np.random.choice(self.noises)
+        noise = None
+        if noise_path is not None:
+            noise_db = np.random.uniform(self.noise_db_low, self.noise_db_high)
+            with soundfile.SoundFile(noise_path) as f:
+                if f.frames == nsamples:
+                    noise = f.read(dtype=np.float64, always_2d=True)
+                elif f.frames < nsamples:
+                    offset = np.random.randint(0, nsamples - f.frames)
+                    # noise: (Time, Nmic)
+                    noise = f.read(dtype=np.float64, always_2d=True)
+                    # Repeat noise
+                    noise = np.pad(
+                        noise,
+                        [(offset, nsamples - f.frames - offset), (0, 0)],
+                        mode="wrap",
```

Comment:
This is nice but could lead to funny noise clips if the noise is very short. Maybe raise a warning ? (but could be too verbose as it is in the dataloader).

Reply:
Thanks for the comment. I have added support for it.